Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -62,78 +164,188 @@ public void Initialize()
-                        while (true)
+                        uint256 hash = this.UTXOSet.Rewind().GetAwaiter().GetResult();
+                        ChainedBlock rewinded = this.Chain.GetBlock(hash);
+                        if (rewinded == null)
                         {
-                            uint256 hash = this.UTXOSet.Rewind().GetAwaiter().GetResult();
-                            ChainedBlock rewinded = this.Chain.GetBlock(hash);
-                            if (rewinded == null)
-                            {
-                                this.logger.LogTrace("Rewound to '{0}', which is still not a part of the current best chain, rewinding further.", hash);
-                                continue;
-                            }
-
-                            this.logger.LogTrace("Rewound to '{0}'.", hash);
-                            this.Tip = rewinded;
-                            this.Puller.SetLocation(rewinded);
-                            break;
+                            this.logger.LogTrace("Rewound to '{0}', which is still not a part of the current best chain, rewinding further.",
+                                hash);
+                            continue;
                         }
+
+                        this.Tip = rewinded;
+                        this.Puller.SetLocation(rewinded);
+                        this.logger.LogInformation("Reorg detected, rewinding from '{0}' to '{1}'.", lastTip, this.Tip);
+                        return;
                     }
                 }
 
-                this.AcceptBlock(new ContextInformation(result, this.Validator.ConsensusParams));
-            }
-            catch (ConsensusErrorException ex)
-            {
-                result.Error = ex.ConsensusError;
+                try
+                {
+                    this.ValidateBlock(new ContextInformation(item, this.Validator.ConsensusParams));
+                }
+                catch (ConsensusErrorException ex)
+                {
+                    item.Error = ex.ConsensusError;
+                }
+
+                if (item.Error != null)
+                {
+                    this.logger.LogError("Block rejected: {0}", item.Error.Message);
+
+                    // Pull again.
+                    this.Puller.SetLocation(this.Tip);
+
+                    if (item.Error == ConsensusErrors.BadWitnessNonceSize)
+                    {
+                        this.logger.LogInformation("You probably need witness information, activating witness requirement for peers.");
+                        this.connectionManager.AddDiscoveredNodesRequirement(NodeServices.NODE_WITNESS);
+                        this.Puller.RequestOptions(TransactionOptions.Witness);
+                        return;
+                    }
+
+                    // Set the chain back to ConsensusLoop.Tip.
+                    this.Chain.SetTip(this.Tip);
```

Comment:
this seems dangerous, but it is probably necessary
is this the improvement that should solve the situation when a peer claims long chain but then we find out that the chain is invalid (previously this would leave the tip at the high value and only after the real chain would advance enough, it would catch up) ?

the little problem with this is that when you are here, you could already see a valid chain that is longer than this.Tip, so you are throwing something valid away, but maybe that is fine to do - my concern would be that your good peers won't resend you those headers and so it seems to me that you can actually be stuck if this happens

i'm not sure if i'm explaining this race condition i have in the mind well, please let me know if you can see it

Reply:
the problem is little bit worse, because let's say we have one bad peer and two good peers, bad peer gives us headers going up to block 200, but real chain is just 150 and we are (consensus) at 100

so this resets the consensus to 100, but those two guys won't send you headers again, so you are still at 100 and i don't think you would even try to get more blocks because you think the chain is at 100

what you could do, however, is to go over all your good peers and check what's their position and set headers there

Reply:
Now thinking about it again, this seems to me that it is actually an improvement over the current state. It is probably true that good peers won't send you headers, so you would be stuck in block 100, but when 1 new block arrives, you will get that information and you will sync to 151. Previously you would need to wait until block 201 is available. So I think this is an improvement and maybe it is good enough and maybe we don't need to go over the list of good peers and set tip to their best.

Reply:
This functionality is same as before but I see your point, it is necessary though.
It will be possible to detect block 150 yeah if we disconnect from the 200 (malicious peer) and assume the rest are honest.

This is should definitely not happen in the consensus code but we may be able to trigger such an operation.

There are 3 ways we can fix this (as part of another PR)
- Just disconnect the peer and don't change Chain.Tip and let issue #229 rewind the header chain to the most recent best height. this will never set chain to 100 but have it directly set to 150 slightly delays.
- Create a new behavior that will do the same as the previous point but triggered from consensus (not from node disconnect).
- Set Chain tip to consensus and trigger out nodes to sync the chain (send an GetHeaders to all peers) this will set Cahin back to 100 and sync it up to 150

Reply:
I think for this PR this is good enough. Further improvements should be out of scope here as you say. So I'm OK here.

Reply:
this is done